server_interfaces_test.go

```diff
@@ -0,0 +1,348 @@
+package dbus
+
+import (
+	"testing"
+)
+
+type tester struct {
+	conn *Conn
+	sigs chan *Signal
+}
+
+type intro struct {
+	path ObjectPath
+}
+
+func (i *intro) introspectPath(path ObjectPath) string {
+	switch path {
+	case "/":
+		return `<node><node name="com"></node></node>`
+	case "/com":
+		return `<node><node name="github"></node></node>`
+	case "/com/github":
+		return `<node><node name="godbus"></node></node>`
+	case "/com/github/godbus":
+		return `<node><node name="tester"></node></node>`
+	}
+	return ""
+}
+
+func (i *intro) LookupInterface(name string) (Interface, bool) {
+	if name == "org.freedesktop.DBus.Introspectable" {
+		return i, true
+	}
+	return nil, false
+}
+
+func (i *intro) LookupMethod(name string) (Method, bool) {
+	if name == "Introspect" {
+		return intro_fn(func() string {
+			return i.introspectPath(i.path)
+		}), true
+	}
+	return nil, false
+}
+
+func newIntro(path ObjectPath) *intro {
+	return &intro{path}
+}
+
+//Handler
+func (t *tester) LookupObject(path ObjectPath) (ServerObject, bool) {
+	if path == "/com/github/godbus/tester" {
+		return t, true
+	}
+	return newIntro(path), true
+}
+
+//ServerObject
+func (t *tester) LookupInterface(name string) (Interface, bool) {
+	switch name {
+	case "com.github.godbus.dbus.Tester":
+		return t, true
+	case "org.freedesktop.DBus.Introspectable":
+		return t, true
+	}
+
+	return nil, false
+}
+
+//Interface
+func (t *tester) LookupMethod(name string) (Method, bool) {
+	switch name {
+	case "Test":
+		return t, true
+	case "Introspect":
+		return intro_fn(func() string {
+			return `<node>
+    <interface name="org.freedesktop.DBus.Introspectable.Introspect">
+        <method name="Introspect">
+            <arg name="out" type="i" direction="out">
+        </method>
+    </interface>
+    <interface name="com.github.godbus.dbus.Tester">
+        <method name="Test">
+            <arg name="in" type="i" direction="in">
+            <arg name="out" type="i" direction="out">
+        </method>
+        <signal name="sig1">
+            <arg name="out" type="i" direction="out">
+        </signal>
+    </interface>
+</node>`
+		}), true
+	}
+	return nil, false
+}
+
+//Method
+func (t *tester) Call(args ...interface{}) ([]interface{}, error) {
+	return args, nil
+}
+
+func (t *tester) NumArguments() int {
+	return 1
+}
+
+func (t *tester) NumReturns() int {
+	return 1
+}
+
+func (t *tester) ArgumentValue(position int) interface{} {
+	return ""
+}
+
+func (t *tester) ReturnValue(position int) interface{} {
+	return ""
+}
+
+//SignalHandler
+func (t *tester) DeliverSignal(iface, name string, signal *Signal) {
+	if iface != "com.github.godbus.dbus.Tester" || name != "sig1" {
```

Comment:
This seems a bit test specific. I understand you have to filter out the noise, but it makes is difficult to use this tester for a different test.